tpm2/direct/commands.go

```diff
@@ -0,0 +1,425 @@
+	PCRSelectionIn TPMLPCRSelection
+}
+
+// Command implements the Command interface.
+func (*PCRReadCommand) Command() TPMCC { return TPMCCPCRRead }
+
+// PCRReadResponse is the response from TPM2_PCR_Read.
+type PCRReadResponse struct {
+	// the current value of the PCR update counter
+	PCRUpdateCounter uint32
+	// the PCR in the returned list
+	PCRSelectionOut TPMLPCRSelection
+	// the contents of the PCR indicated in pcrSelectOut-> pcrSelection[] as tagged digests
+	PCRValues TPMLDigest
+}
+
+// Response implements the Response interface.
+func (*PCRReadResponse) Response() TPMCC { return TPMCCPCRRead }
+
+// PolicySecretCommand is the input to TPM2_PolicySecret.
+// See definition in Part 3, Commands, section 23.4
+type PolicySecretCommand struct {
+	// handle for an entity providing the authorization
+	AuthHandle AuthHandle `gotpm:"handle,auth"`
+	// handle for the policy session being extended
+	PolicySession TPMISHPolicy `gotpm:"handle"`
+	// the policy nonce for the session
+	NonceTPM TPM2BNonce
+	// digest of the command parameters to which this authorization is limited
+	CPHashA TPM2BDigest
+	// a reference to a policy relating to the authorization – may be the Empty Buffer
+	PolicyRef TPM2BNonce
+	// time when authorization will expire, measured in seconds from the time
+	// that nonceTPM was generated
+	Expiration int32
+}
+
+// Command implements the Command interface.
+func (*PolicySecretCommand) Command() TPMCC { return TPMCCPolicySecret }
+
+// PolicySecretResponse is the response from TPM2_PolicySecret.
+type PolicySecretResponse struct {
+	// implementation-specific time value used to indicate to the TPM when the ticket expires
+	Timeout TPM2BTimeout
+	// produced if the command succeeds and expiration in the command was non-zero
+	PolicyTicket TPMTTKAuth
+}
+
+// Response implements the Response interface.
+func (*PolicySecretResponse) Response() TPMCC { return TPMCCPolicySecret }
+
+// CreatePrimaryCommand is the input to TPM2_CreatePrimary.
+// See definition in Part 3, Commands, section 24.1
+type CreatePrimaryCommand struct {
+	// TPM_RH_ENDORSEMENT, TPM_RH_OWNER, TPM_RH_PLATFORM+{PP},
+	// or TPM_RH_NULL
+	PrimaryHandle AuthHandle `gotpm:"handle,auth"`
+	// the sensitive data
+	InSensitive TPM2BSensitiveCreate
+	// the public template
+	InPublic TPM2BPublic
+	// data that will be included in the creation data for this
+	// object to provide permanent, verifiable linkage between this
+	// object and some object owner data
+	OutsideInfo TPM2BData
+	// PCR that will be used in creation data
+	CreationPCR TPMLPCRSelection
+}
+
+// Command implements the Command interface.
+func (*CreatePrimaryCommand) Command() TPMCC { return TPMCCCreatePrimary }
+
+// CreatePrimaryResponse is the response from TPM2_CreatePrimary.
+type CreatePrimaryResponse struct {
+	// handle of type TPM_HT_TRANSIENT for created Primary Object
+	ObjectHandle TPMHandle `gotpm:"handle"`
+	// the public portion of the created object
+	OutPublic TPM2BPublic
```

Comment:
We can consider collapsing TPM2B with the thing that it wraps, annotating it here with the "sized" instead and eliminating the 2B types.

This allows us to also rename stuff, eg `TPMHandle` as `Handle` and avoid the stuttering of `tpm2.TPMHandle`.

Reply:
I thought about this. I don't want to collapse TPM2B because a user looking at the spec cares about the difference between a TPM2BSomething and the TPMSSomething it wraps (i.e., they might be serializing/deserializing one or the other).

We can fix the stuttering with another suggestion that came up during our discussion, which is to provide subpackages for all these prefixes.